messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ActiveMQResourceAdapter.java

```diff
@@ -79,4 +85,72 @@ protected BroadcastEndpointFactory createBroadcastEndpointFactory(ConnectionFact
     private static ServiceContainer currentServiceContainer() {
         return (System.getSecurityManager() == null) ? CurrentServiceContainer.getServiceContainer() : AccessController.doPrivileged(CurrentServiceContainer.GET_ACTION);
     }
+
+    @Override
+    public WorkManager getWorkManager() {
+        return new NamingWorkManager(super.getWorkManager());
+    }
+
+    private class NamingWorkManager implements WorkManager {
+
+        private final WorkManager delegate;
+
+        private NamingWorkManager(WorkManager delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void doWork(Work work) throws WorkException {
+            delegate.doWork(work);
+        }
+
+        @Override
+        public void doWork(Work work, long startTimeout, ExecutionContext execContext, WorkListener workListener) throws WorkException {
+            delegate.doWork(work, startTimeout, execContext, workListener);
+        }
+
+        @Override
+        public long startWork(Work work) throws WorkException {
+            return delegate.startWork(work);
+        }
+
+        @Override
+        public long startWork(Work work, long startTimeout, ExecutionContext execContext, WorkListener workListener) throws WorkException {
+            return delegate.startWork(work, startTimeout, execContext, workListener);
+        }
+
+        @Override
+        public void scheduleWork(Work work) throws WorkException {
+            Work myWork = new WorkWrapper(NamespaceContextSelector.getCurrentSelector(), work);
+            delegate.scheduleWork(myWork);
+        }
+
+        @Override
+        public void scheduleWork(Work work, long startTimeout, ExecutionContext execContext, WorkListener workListener) throws WorkException {
+            delegate.scheduleWork(work, startTimeout, execContext, workListener);
+        }
+    }
+
+    private class WorkWrapper implements Work {
+
+        private final Work delegate;
+        private final NamespaceContextSelector selector;
+
+        private WorkWrapper(NamespaceContextSelector selector, Work work) {
+            this.selector = selector;
+            this.delegate = work;
+        }
+
+        @Override
+        public void release() {
+            delegate.release();
+            NamespaceContextSelector.popCurrentSelector();
+        }
+
+        @Override
+        public void run() {
```

Comment:
release() seems to be just a cancel() but just to be safe, the selector should be pushed and popped in both methods, using try blocks around the delegation.